### PR TITLE
Inconsistent Model Path

### PR DESCRIPTION
**Pull request by dpanshug**
[RHOAIENG-542](https://issues.redhat.com/browse/RHOAIENG-542)

## Description
Removes the leading slash `/` single or multiple and throws error msg for invalid path

[Screencast from 2024-01-30 14-49-34.webm](https://github.com/opendatahub-io/odh-dashboard/assets/97534722/32451a96-cf75-407c-a12d-b38f0e85ab80)

Invalid path

Screenshot from 2024-01-30 14-47-38

Error when only slash `/` in path 

Screenshot from 2024-01-30 14-47-30

## How Has This Been Tested?
1. Deploy new model KServe or Multi model
2. Enter path name starting with '/'
3. Edit the model created, you can see the '/' removed from path.

## Test Impact
Added unit test for utility function.

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 

- [x] Included any necessary screenshots or gifs if it was a UI change.
- [x] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

@vconzola

### CODE REVIEW

**Comment by dpanshug** [comment]
@vconzola @kaedward pls check the error msg.

**Comment by vconzola** [comment]
@andrewballantyne @lucferbux @DaoDaoNoCode i thought we addressed this a looong time ago by just ignoring any leading "/" or "/////" that a user typed in the path field?

**Comment by manaswinidas** [comment]
@vconzola Is this what you are talking about https://github.com/opendatahub-io/odh-dashboard/pull/1512

**Review by manaswinidas — COMMENTED** [review]
I think this issue is more about whether the backend needs us to prefix or not. And whether the prefix is coming from the backend or from us? please correct me if I'm missing something here. I remember this being discussed in one of the Triage calls.

**Comment by dpanshug** [comment]
@vconzola This one is different because it introduces additional functionality to raise an error when characters appear after '/'. In the previous version, errors were only triggered for the '/' character itself. Now, errors will also be raised for paths like '/something'."

So for path 
1. '/' error is "The path must not point to a root folder."
2. '/something' error is "The path must not start with /"

**Comment by lucferbux** [comment]
> > @vconzola @kaedward pls check the error msg.
> 
> How about "Must start with a letter or number"? Can paths start with numbers or other characters, or only letters?

That's a great question, and something that we are not checking cc @Jooho @Xaenalt can you guys help us out here to validate that too?

**Comment by lucferbux** [comment]
> @andrewballantyne @lucferbux @DaoDaoNoCode i thought we addressed this a looong time ago by just ignoring any leading "/" or "/////" that a user typed in the path field?

It seems that the logic behind it wasn't working well and there was an issue, that's why we had [RHOAIENG-542](https://issues.redhat.com/browse/RHOAIENG-542) in place. But checking a solution for it we realized that we shouldn't allow a leading "/" in any case, so the proposed solution would be error out that rather than silently remove it, what do you think?

**Comment by vconzola** [comment]
I'm still not sure I understand the issue completely, but in general the more "fault tolerant" we can be for users, the better. So I'm not sure why just ignoring any leading "/" won't work? Regarding [RHOAIENG-542](https://issues.redhat.com/browse/RHOAIENG-542) (if I understand that issue correctly), we should not store any leading "/" so when the path is rendered for editing there will be no leading "/".

**Comment by dpanshug** [comment]
> So I'm not sure why just ignoring any leading "/" won't work? 

This PR introduces a condition specifically targeting chars following a "/" in a path. Because, previously the error would only be triggered if the path was exactly equal to "/".

> We would still throw an error ("The path must not point to a root folder.") if the user leaves the field blank or inputs only one or more "/" to address the root folder issue.

Yes I also think we should keep the error msg same.

**Comment by vconzola** [comment]
> For example, what I'm understanding is the following:
> 
> / -> Will fail ❌
> path/model -> Will work and it's correct ✅
> /path/model -> Will work, but we should filter the leading slash and convert it to path/model ⚠️
> 
> Is that right?

@lucferbux Yes, that's what I suggest (assuming it's technically correct).

**Comment by dpanshug** [comment]
@lucferbux @vconzola made necessary changes :heavy_check_mark:

**Comment by lucferbux** [comment]
Ok, lat nit, and I think it should be addressed here:

Screenshot 2024-01-29 at 12 41 51

Trying this won't let you input the model, but the message is wrong, since it's not pointing at the root folder, this happens cause we have this in the validation:

`if (folderPath === '/' || folderPath.includes('//'))`

We should do two small changes:

1. Change the validation to test wether it includes ONLY one or several slashes (/ or //// or //)
2. Change your filter method to filter ALL the leading slashes

File: frontend/src/__tests__/cypress/cypress/pages/modelServing.ts [hunk 1]
```diff
@@ -115,6 +115,10 @@ class InferenceServiceModal extends Modal {
   findLocationPathInput() {
     return this.find().findByLabelText('folder-path');
   }
+
+  findLocationPathInputError() {
+    return this.find().findByLabelText('folder-path-error');
```
Comment (manaswinidas) [inline frontend/src/__tests__/cypress/cypress/pages/modelServing.ts:120]: I think this should use `test-id`.

**Review by manaswinidas — CHANGES_REQUESTED** [review]

**Review by manaswinidas — APPROVED** [review]
/lgtm

**Review by lucferbux — APPROVED** [review]
/lgtm

Great job!!

**Pull request merged**